File: src/structures/bst.cpp
// bst.cpp — CLRS 4e Ch12：BST 插入／搜尋／走訪／刪除（transplant 版）
#include <iostream>

struct Node {
    int key;
    Node *left = nullptr, *right = nullptr, *p = nullptr;
    explicit Node(int k) : key(k) {}
};

struct BST {
    Node* root = nullptr;

    void insert(int k) {                       // TREE-INSERT：沿途二選一往下走
        Node* z = new Node(k);
        Node *y = nullptr, *x = root;
        while (x) { y = x; x = (k < x->key) ? x->left : x->right; }
        z->p = y;
        if (!y) root = z;
        else if (k < y->key) y->left = z;
        else y->right = z;
    }
    Node* search(Node* x, int k) const {
        while (x && x->key != k) x = (k < x->key) ? x->left : x->right;
        return x;
    }
    Node* minimum(Node* x) const { while (x->left) x = x->left; return x; }

    void transplant(Node* u, Node* v) {        // 用 v 的子樹取代 u 的位置
        if (!u->p) root = v;
        else if (u == u->p->left) u->p->left = v;
        else u->p->right = v;
        if (v) v->p = u->p;
    }
    void erase(Node* z) {                      // TREE-DELETE 三情境
        if (!z->left) transplant(z, z->right);
        else if (!z->right) transplant(z, z->left);
        else {
            Node* y = minimum(z->right);       // 後繼：右子樹最小
            if (y->p != z) {
                transplant(y, y->right);
                y->right = z->right; y->right->p = y;
            }
            transplant(z, y);
            y->left = z->left; y->left->p = y;
        }
        delete z;
    }
    void inorder(Node* x) const {
        if (!x) return;
        inorder(x->left);
        std::cout << x->key << ' ';
        inorder(x->right);
    }
};

int main() {
    BST T;
    for (int k : {12, 5, 18, 2, 9, 15, 19, 17}) T.insert(k);
    T.inorder(T.root); std::cout << '\n';
    T.erase(T.search(T.root, 12));             // 刪除有兩個子節點的根
    T.inorder(T.root); std::cout << '\n';
    std::cout << "min = " << T.minimum(T.root)->key << '\n';
    return 0;
}

File: src/structures/bst.py
# 改寫自官方 Chapter 12/binary_search_tree.py（MIT License）：Node＋平鋪函式
class Node:
    def __init__(self, key):
        self.key = key
        self.left = self.right = self.p = None


def tree_insert(T, z):
    x, y = T["root"], None
    while x is not None:                 # 追蹤落點
        y = x
        x = x.left if z.key < x.key else x.right
    z.p = y
    if y is None:
        T["root"] = z
    elif z.key < y.key:
        y.left = z
    else:
        y.right = z


def inorder(x, out):
    if x is not None:
        inorder(x.left, out)
        out.append(x.key)
        inorder(x.right, out)


def tree_minimum(x):
    while x.left is not None:
        x = x.left
    return x


def transplant(T, u, v):                 # 用子樹 v 整棵取代 u
    if u.p is None:
        T["root"] = v
    elif u is u.p.left:
        u.p.left = v
    else:
        u.p.right = v
    if v is not None:
        v.p = u.p


def tree_delete(T, z):
    if z.left is None:                   # 情境①：無左子 → 右子頂上
        transplant(T, z, z.right)
    elif z.right is None:                # 情境②：無右子 → 左子頂上
        transplant(T, z, z.left)
    else:                                # 情境③：兩子俱在 → 後繼接手
        y = tree_minimum(z.right)        # 後繼必無左子
        if y.p is not z:
            transplant(T, y, y.right)    # y 先脫離原位
            y.right = z.right
            y.right.p = y
        transplant(T, z, y)
        y.left = z.left
        y.left.p = y


def search(x, k):
    while x is not None and x.key != k:
        x = x.left if k < x.key else x.right
    return x


T = {"root": None}
for k in [12, 5, 18, 2, 9, 15, 19, 17]:
    tree_insert(T, Node(k))
out = []
inorder(T["root"], out)
print("inorder:", out)
tree_delete(T, search(T["root"], 12))    # 刪根（兩子俱在 → 後繼 15 接手）
out = []
inorder(T["root"], out)
print("刪 12 後:", out, "，新根 =", T["root"].key)
